### PR TITLE
Implement process_counts in SampleMP

### LINKED ISSUE #5371: Implement `SampleMP.proces_counts`
# Context

Currently, our sample-based measurement processes that inherit from `SampleMeasurement` implement a [process_samples](https://github.com/PennyLaneAI/pennylane/blob/a05ae286c5925e0c881099cea751193d368779fc/pennylane/measurements/measurements.py#L535) method. This method is capable of turning any samples (array of int64) into the required measurement result. This method is currently used internally by `default.qubit`.

```
@qml.qnode(qml.device('default.qubit'))
def circuit():
    qml.Hadamard(0)
    return qml.sample(wires=(0,1))

samples = circuit(shots=10)
qml.sample(wires=0).process_samples(samples, wire_order=qml.wires.Wires((0,1)))
```

But external devices and plugins may want to store sampling information in the form of a counts dictionary instead, like `{'00': 45, '01': 55}`. This representation is much more condensed and memory efficient. Because of this, we have recently added a `SampleMeasurement.process_counts` method.

# Implementation Details

The task is to implement [SampleMP.process_counts(counts, wire_order)](https://github.com/PennyLaneAI/pennylane/blob/3407a9e3fd76f0b112ef22e14ef8a4ba3d5a339f/pennylane/measurements/counts.py#L159) in such a way that a counts dictionary to look like raw samples.

```pycon
>>> counts = {'00': 2, '10': 3}
>>> qml.sample(wires=(0,1)).process_counts(counts, wire_order=qml.wires.Wires((0,1)))
array([[0, 0],
       [0, 0],
       [1, 0],
       [1, 0],
       [1, 0]])
>>> qml.sample(wires=0).process_counts(counts, wire_order=qml.wires.Wires((0,1)))
array([0, 0, 1, 1, 1])
```

# Requirements

A completed PR should:

* Add the `process_counts` method to `SampleMP`
* Add unit tests to [pennylane/tests/measurements/test_sample.py](https://github.com/PennyLaneAI/pennylane/blob/master/tests/measurements/test_sample.py)
* Pass black and pylint checks
* Add a changelog entry to [pennylane/releases/changelog-dev.md](https://github.com/PennyLaneAI/pennylane/blob/master/doc/releases/changelog-dev.md) under Improvements. Don't forget to add your name to the list of contributors at the bottom :)

### PR DESCRIPTION
**Pull request by Tarun-Kumar07**
**Context:**
Under #4941 abstract method `process_counts` was added to`SampleMeasurement`. This method provides support to process counts dictionary produced by external systems.

**Description of the Change:**
* Implement `process_counts` in `SampleMP`.
* Make `SampleMeasurement.process_counts` an abstract method

**Benefits:**
The method `SampleMP.process_counts` won't throw `NotImplementedError` exception

**Possible Drawbacks:**

**Related GitHub Issues:**
Fixes #5371

### CODE REVIEW

**Comment by Tarun-Kumar07** [comment]
@albi3ro, I have a question regarding the method `SampleMeasurement.process_counts`.  
Would it be possible to `make it abstract` now? 
The reason for this is that all the child classes inheriting from it in the main source code have already implemented this method.

**Comment by astralcai** [comment]
> @albi3ro, I have a question regarding the method `SampleMeasurement.process_counts`. Would it be possible to `make it abstract` now? The reason for this is that all the child classes inheriting from it in the main source code have already implemented this method.

Isn't `SampleMeasurement.process_counts` already abstract? I see that it raises `NotImplementedError`

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -159,6 +159,9 @@
 * Implemented the method `process_counts` in `ExpectationMP`, `VarianceMP`, and `CountsMP`.
   [(#5256)](https://github.com/PennyLaneAI/pennylane/pull/5256)
 
+* Implemented the method `process_counts` in `SampleMP`
+  [(#5395)](https://github.com/PennyLaneAI/pennylane/pull/5395)
+
```
Comment (astralcai) [inline doc/releases/changelog-dev.md:164]: You can probably combine this entry with the one above.

```suggestion
* Implemented the method `process_counts` in `ExpectationMP`, `VarianceMP`, `CountsMP`, and `SampleMP`
  [(#5256)](https://github.com/PennyLaneAI/pennylane/pull/5256)
  [(#5395)](https://github.com/PennyLaneAI/pennylane/pull/5395)

```

**Comment by albi3ro** [comment]
> > @albi3ro, I have a question regarding the method `SampleMeasurement.process_counts`. Would it be possible to `make it abstract` now? The reason for this is that all the child classes inheriting from it in the main source code have already implemented this method.
> 
> Isn't `SampleMeasurement.process_counts` already abstract? I see that it raises `NotImplementedError`

By abstract, we mean decorated with `@abc.abstractmethod`.  This *requires* every child of `SampleMeasurement` to define `process_counts`, and you can't create a `SampleMeasurement` without overriding the method.

I'm on the fence about making it abstract or not.  On one hand, everything in PennyLane defines it right now, and we want to make sure it keeps being defined in the future.  On the other side of things, this method isn't actually *used* by any workflows I'm aware of right now, so it's not quite as important.

Let's go ahead and make it abstract like `process_samples`.  I can't really see it causing that many problems, and it will enforce better feature coverage in the future.

**Review by astralcai — APPROVED** [review]
Looks good to me! 🎸

**Comment by Tarun-Kumar07** [comment]
@astralcai I was suggesting the following change, similar to `process_samples`
```python
    @abstractmethod
    def process_counts(self, counts: dict, wire_order: Wires):
         ...
```

This ensure that whenever some <u>child class</u> of `SampleMeasurement` are developed in future,
They will fail at instantiation with `TypeError`. 
Now they fail when method `process_counts` is called.

This would also need alignments in some test classes where there are child classes of `SampleMeasurement`, but no change in the main source code.

File: tests/measurements/test_sample.py
Comment (albi3ro) [inline tests/measurements/test_sample.py:1]: Potentially worth also having tests for:

1) sampling on the second wire: `qml.sample(wires=1)`, `wire_order=qml.wires.Wires((0,1))`
2) inverting the wire order: `qml.sample(wires=("b", "a"))`, `wire_order=qml.wires.Wires(("b", "a"))`

File: pennylane/devices/tests/test_measurements.py [hunk 2]
```diff
@@ -1630,6 +1631,9 @@ class MyMeasurement(SampleMeasurement):
             def process_samples(self, samples, wire_order, shot_range=None, bin_size=None):
                 return 1
 
+            def process_counts(self, counts: dict, wire_order: Wires):
+                return counts
```
Comment (albi3ro) [inline pennylane/devices/tests/test_measurements.py:1635]:
```suggestion
                return 1
```

Just makes it compatible with the `process_samples` method.

File: pennylane/devices/tests/test_measurements.py [hunk 3]
```diff
@@ -1651,6 +1655,9 @@ class MyMeasurement(SampleMeasurement):
...
```
Comment (albi3ro) [inline pennylane/devices/tests/test_measurements.py:1659]:
```suggestion
                return 1
```

**Review by albi3ro — APPROVED** [review]
Looks great!

We're currently having some CI issues due to an updated lightning device, but I'll mark this as `merge-ready`, and we'll merge it in once the various issues are resolved.

**Comment by albi3ro** [comment]
I know that code cov was passing, but it is just hanging due to having to rerun a test with a stochastic failure.  So I'm overriding and merging.

**Pull request merged**